src/featured_graph.jl

```diff
@@ -0,0 +1,278 @@
+#===================================
+Define FeaturedGraph type as a subtype of LightGraphs' AbstractGraph.
+For the core methods to be implemented by any AbstractGraph, see
+https://juliagraphs.org/LightGraphs.jl/latest/types/#AbstractGraph-Type
+https://juliagraphs.org/LightGraphs.jl/latest/developing/#Developing-Alternate-Graph-Types
+=============================================#
+
+abstract type AbstractFeaturedGraph <: AbstractGraph{Int} end
+
+"""
+    NullGraph()
+
+Null object for `FeaturedGraph`.
+"""
+struct NullGraph <: AbstractFeaturedGraph end
+
+
+struct FeaturedGraph <: AbstractFeaturedGraph
+    edge_index
+    num_nodes::Int
+    num_edges::Int
+    # ndata::Dict{String, Any} # https://github.com/FluxML/Zygote.jl/issues/717        
+    # edata::Dict{String, Any}
+    # gdata::Dict{String, Any}
+    nf
+    ef
+    gf
+end
+
+
+function FeaturedGraph(s::AbstractVector{Int}, t::AbstractVector{Int}; 
+                        num_nodes = max(maximum(s), maximum(t)), 
+                        # ndata = Dict{String, Any}(), 
+                        # edata = Dict{String, Any}(),
+                        # gdata = Dict{String, Any}(),
+                        nf = nothing, 
+                        ef = nothing, 
+                        gf = nothing)
+
+    @assert length(s) == length(t)
+    @assert min(minimum(s), minimum(t)) >= 1 
+    @assert max(maximum(s), maximum(t)) <= num_nodes 
+    
+    num_edges = length(s)
+    
+    ## I would like to have dict data store, but currently this 
+    ## doesn't play well with zygote due to 
+    ## https://github.com/FluxML/Zygote.jl/issues/717    
+    # ndata["x"] = nf
+    # edata["e"] = ef
+    # gdata["g"] = gf
+    
+
+    FeaturedGraph((s, t), num_nodes, num_edges, 
+                   nf, ef, gf)
+end
+
+# Construct from adjacency matrix # TODO deprecate?
+function FeaturedGraph(adj_mat::AbstractMatrix; dir=:out, kws...)
+    @assert dir == :out  # TODO
+    num_nodes = size(adj_mat, 1)
+    @assert num_nodes == size(adj_mat, 2)
+    @assert all(x -> (x == 1) || (x == 0), adj_mat)
+    num_edges = round(Int, sum(adj_mat))
+    s = zeros(Int, num_edges)
+    t = zeros(Int, num_edges)
+    e = 0
+    for j in 1:num_nodes
+        for i in 1:num_nodes
+            if adj_mat[i, j] == 1
+                e += 1
+                s[e] = i
+                t[e] = j
+            end
+        end
+    end
+    @assert e == num_edges
+    FeaturedGraph(s, t; num_nodes, kws...)
+end
+
+
+# Construct from adjacency list # TODO deprecate?
```

Comment:
So, what do you think of a best data structure to construct from?

Reply:
actually we want to support construction out of any graph representation, I'll remove this comment